source/installation_guide/upgrading/from-2.3-to-3.0.rst

```diff
@@ -92,3 +93,82 @@ Removed features and their replacements
 +------------------------------------------------------------+------------------------------------------------------------------------------------------+
 | checkpassword auth database                                | Use :ref:`authentication-lua_based_authentication` instead.                              |
 +------------------------------------------------------------+------------------------------------------------------------------------------------------+
+
+Use Global ACL Files instead of Global ACL Directories
+------------------------------------------------------
+
+To migrate the ACL directories into their respective files you have to do the
+following:
+
+#. create a new consolidated :ref:`acl-global_acl_file`,
+#. for each subdirectory in the currently configured ACL directory add a line
+   starting with the mailbox name (i.e. the path from the ACL directory parent)
+   followed by the appropriate content,
+#. change the vfile parameter to the new ACL file, and finally
+#. remove the old ACL directory parent.
+
+Example
+^^^^^^^
+
+With the following starting configuration:
+
+.. code-block:: none
+
+   # dovecot.conf
+
+   plugin {
+     acl = vfile:/etc/dovecot/acls/
+   }
+
+.. code-block:: none
+
+   # /etc/dovecot/acls/INBOX
+
+   owner lrwstipekxa
+   anyone lr
+   user=kim l
+
+.. code-block:: none
+
+   # /etc/dovecot/acls/foo/.DEFAULT
+
+   user=timo lr
+   user=kim lrw
+
+.. code-block:: none
+
+   # /etc/dovecot/acls/foo/bar
+
+   user=kim lrw
+
+You have to create the new ACL file:
+
+.. code-block:: none
+
+   # /etc/dovecot/dovecot-acl
+
+   # previously from /etc/dovecot/acls/INBOX
+   INBOX owner lrwstipekxa
+   INBOX anyone lr
+   INBOX user=kim l
+   # previously from /etc/dovecot/acls/foo/.DEFAULT
+   foo user=timo lr
+   foo user=kim lrw
+   # previously from /etc/dovecot/acls/foo/bar
+   foo/bar user=kim lrw
+
```

Comment:
This should also talk about the naming differences between ACL directory and ACL file, and give examples. A good example could be using namespace { prefix=INBOX/ }. With ACL directory it's using "storage name" while with ACL file it's using "virtual name". It has two big differences: virtual name includes the namespace prefix, and uses the namespace's hierarchy separator. Storage name doesn't include namespace prefix, and uses the mailbox format's internal hierarchy separator. You could link to "Mailbox names" in source/developer_manual/design/mailbox_list.rst

Reply:
OK, I think I know what you mean and amended the commit.

But https://doc.dovecot.org/configuration_manual/acl/#global-acl-directory-obsolete explicitly states:

> They are looked up using the mailbox’s virtual name.

and

> The filenames must start with namespace prefix (if it has one).

Which seems to be the exact opposite of what you mentioned regarding the "storage name" format.

Reply:
Oops.. I was looking at acl_backend_vfile_has_acl() where it was using storage name, but looks like acl_backend_vfile_has_acl() was just buggy. For the most part, the global acl directory really was using vnames. So, sorry, please revert. :) But having namespace prefix in the examples is still a good idea anyway.